src/schedule/ll_schedule.c

```diff
@@ -171,10 +171,16 @@ static void schedule_ll_tasks_run(void *data)
 
 	perf_cnt_init(&sch->pcd);
 
+	notifier_event(sch, NOTIFIER_ID_LL_PRE_RUN,
```

Comment:
Why are we doing this in the scheduler ?

Reply:
It can be done once for all DMA channels, so no point in doing it after every DMA copy separately. We're just wasting cycles. With timer driven pipelines we are sure that it will be executed only once for all channels.

Reply:
Ok, but I'm not following the usage here of L1 exit and when we need to use it. i.e. does it apply to all channels together or individually, do we exit L1 after DMA has stopped etc. Can you outline the rules a little more. Thanks

Reply:
It applies to all channels on all cores together and we should exit L1 after DMA pointer increment. The usage is exactly the same as before, but now it only happens once after all pipeline_copy tasks has been executed and not individually after every DMA channel copy.

Reply:
This PM L1 entry/exit should be handled directly by the HDA-DMA driver and not via LL scheduler state changes (as these can be done for any task unrelated to HDA DMA state). What's stopping us running an unrelated task via the LL scheduler during HDA work when we need to be in L1.
I'm also not following why we can't reference count pm_runtime_get(L1)/put(L1) within the HDA code ?

Reply:
There is no HDA-DMA related code directly in the LL scheduler. Everything is done via notification system. 

> What's stopping us running an unrelated task via the LL scheduler during HDA work when we need to be in L1.
LL scheduler works synchronously, there is no possibility of running simultaneous tasks. Also there is no special requirement "to be in L1". We are simply trying to opportunistically exit L1, when there is no other work to do.

Reply:
Sorry, I dont see how we can avoid the following
1) pipeline is doing copy work whilst in L1
2) Some other IRQ schedules some new LL work
3) New LL work completes.
4) Notifier is run.
5) HDA leaves L1 whilst copy work still in progress.

Reply:
Two things:
1. There is no possibility of preemption low latency tasks. It's by definition low latency.
2. Exiting L1 is synchronous:
```
shim_write(SHIM_SVCFG,  shim_read(SHIM_SVCFG) | SHIM_SVCFG_FORCE_L1_EXIT);
wait_delay(PLATFORM_FORCE_L1_EXIT_TIME);
shim_write(SHIM_SVCFG, shim_read(SHIM_SVCFG) & ~(SHIM_SVCFG_FORCE_L1_EXIT));
```

Reply:
But now we have a dependency between **any** LL work and HDA L1 entry. This is broken.

Reply:
@lgirdwood Not true. If there is no LL work that uses HDA DMA, there will be no L1 entry/exit.

Reply:
But wont any LL work invoke the notifier? This is regardless of HDA state, it's then a race between HDA pipelines and any LL work.

Reply:
They are invoked before processing list of LL tasks and after. Not before and after every single task.

Reply:
ok, so my question is still the same, but rather than after/before every LL task, it's running the LL scheduler. Wont this put restrictions on when the scheduler can be used ? i.e. code that wants to schedule work on the LL scheduler will have to know about the state of any pipelines with HDA DMA ?

Reply:
We are keeping ref count, so no special knowledge is needed. It should happen automatically.